src/rust/catmem/queue.rs

```diff
@@ -5,68 +5,292 @@
-    pub fn get_pipe(&self) -> &Pipe {
-        &self.pipe
+    /// Prepares a transition to the closed state.
+    pub fn prepare_closed(&self) -> Result<(), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.prepare_closed(),
+            Ring::PopOnly(ring) => ring.prepare_closed(),
+        }
     }
 
-    /// Set underlying uni-directional pipe.
-    pub fn get_mut_pipe(&mut self) -> &mut Pipe {
-        &mut self.pipe
+    /// This function commits the queue to closing.
+    pub fn commit(&self) {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.commit(),
+            Ring::PopOnly(ring) => ring.commit(),
+        }
+    }
+
+    /// This function aborts a close operation.
+    pub fn abort(&self) {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.abort(),
+            Ring::PopOnly(_) => warn!("abort() called on a pop-only queue"),
+        }
+    }
+
+    fn try_close(&self) -> Result<(), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(ring) => ring.try_close(),
+            Ring::PopOnly(_) => Ok(()),
+        }
+    }
+
+    /// This function
+    pub async fn async_close(&self, yielder: Yielder) -> Result<(), Fail> {
+        for _ in 0..MAX_RETRIES_PUSH_EOF {
+            if let Err(_) = self.try_close() {
+                // We did not, thus retry.
+                match yielder.yield_once().await {
+                    Ok(()) => continue,
+                    Err(cause) => return Err(cause),
+                }
+            } else {
+                return Ok(());
+            }
+        }
+
+        // We ran out of retries, thus fail.
+        let cause: String = format!("failed to push EoF");
+        error!("push_eof(): {}", cause);
+        Err(Fail::new(libc::EIO, &cause))
+    }
+
+    /// This private function tries to pop from the queue and is mostly used for scoping the borrow.
+    fn try_pop(&self) -> Result<(Option<u8>, bool), Fail> {
+        match &mut *self.ring.borrow_mut() {
+            Ring::PushOnly(_) => {
+                let cause: &String = &format!("Cannot pop from push-only queue");
+                error!("{}", &cause);
+                Err(Fail::new(libc::EINVAL, cause))
+            },
+            Ring::PopOnly(ring) => Ok(ring.try_pop()?),
+        }
+    }
+
+    /// This function pops a buffer of optional [size] from the queue. If the queue is connected to the push end of a
+    /// shared memory ring, this function returns an error.
+    pub async fn pop(&self, size: Option<usize>, yielder: Yielder) -> Result<(DemiBuffer, bool), Fail> {
+        let size: usize = size.unwrap_or(limits::RECVBUF_SIZE_MAX);
+        let mut buf: DemiBuffer = DemiBuffer::new(size as u16);
+        let mut index: usize = 0;
+        let eof: bool = loop {
+            match self.try_pop()? {
+                (Some(byte), eof) => {
+                    if eof {
```

Comment:
Not that it must be fixed now, but the match blocks are long, *maybe* segregating into functions might make them simpler to reason about.

Reply:
In some circustances I agree with that statement, but not with this one. Rationale that goes beyond opinions:
- The funtion is small, which according to our guidelines is whatever it fits in one screen.
- Match statements work with pattern matching and enable aggressive inlining.
- This code is performance criticial and agreessive inlining is required.

Reply:
> You can annotate the functions with the inline keyword if necessary. 

- Using inline annotations does not ensure that the compiler will inline. It is a compiler hint. It is true for C code and it is true for rust as well.

> The nested indentation is a bigger concern for me than the actual length here.

Unfortunately Rust is a curly-braced and heavily idented language. 

> Btw, if this is performance critical code, we should measure it using benchmarks and decide based on actual performance numbers.

- No need to measure `pop()` to decide if it is performance critical: pop is on the datapath, and that is a primary concern for demikernel.

Reply:
I meant we should decide which approach is better by measuring (not whether this is a critical path).

Also, if readability/maintainability comes at the cost of acceptable performance hit, we should take it.

Feel free to defer/ignore this suggestion if you are sure that refactoring this into functions will cause unacceptable performance hit.

Reply:
> I meant we should decide which approach is better by measuring (not whether this is a critical path).

Sure, but this is equivalent about measuring time spent on a context switch routine to decide whether or not it is performance critical.

> Indentation: It's not about rust or braces. Generally, multiple indentation levels indicate room for simplification. (Again, this is not critical because it's not common at other places in the overall codebase - just something to keep watch on because future modifications of this code will make complex and possibly add more logic/conditions).

It is, right? Because that is te pattern matching code adding an extra level of indentation there.

> Feel free to defer/ignore this suggestion if you are sure that refactoring this into functions will cause unacceptable performance hit.

Yeah, I would take these modifications if and only if another reviewer asks for the same thing.

Reply:
I'm ok with the current code, so I already approved the PR.

**Benchmarking**
Measuring would be the same in either case, but the decision that I'm looking for is which code is better in both aspects (readability + acceptable performance) - specifically the existing code vs extracting into functions.

**Indentation**
If not pattern matching, you'd use if/else. So that wouldn't make a difference in the indentation. Sure, if there's a way to avoid using match and/or reduce the indentation, it would be interesting to see that variant.